docs/Add-Models.md

```diff
@@ -0,0 +1,270 @@
+
+To convert your config class to and from Hugging Face config class, you can write class functions `to_hf_config()` and `from_hf_config()` that maps the input parameters to the corresponding Hugging Face parameters. For example, in Llama, we have the following:
+
+```python
+from transformers import PretrainedConfig as HfPretrainedConfig
+
+# ...
+
+@classmethod
+def from_hf_config(cls, hf_config: HfPretrainedConfig):
+    return LlamaConfig(
+        seq_len=hf_config.max_position_embeddings,
+        hidden_dim=hf_config.hidden_size,
+        intermediate_dim=hf_config.intermediate_size,
+        num_layers=hf_config.num_hidden_layers,
+        num_heads=hf_config.num_attention_heads,
+        activation_function=hf_config.hidden_act,
+        initializer_range=hf_config.initializer_range,
+        layer_norm_epsilon=hf_config.rms_norm_eps,
+        rope_scaling=hf_config.rope_scaling,
+    )
+```
+
+Lastly, you should register your head model's class name as a class property. This step can be deferred until the head class is constructed.
+This class property would make it easier to call your model with the config class. For example:
+
+```python
+@property
+def model_type(cls) -> Type["LlamaLMHeadModel"]:
+    return LlamaLMHeadModel
+```
+
+### Write Model
+After you have defined your config class, you can start writing your model.
+You can follow the breakdown of the model in your Hugging Face implmentation. This would make it easier to validate the implementation through unit tests.
+
+For example, in GPT2, we have the following breakdown:
+- `Gpt2Mlp`
+- `Gpt2Attention`
+- `Gpt2Block`: a block of Gpt2Mlp and Gpt2Attention
+- `Gpt2Transformer`: a stack of Gpt2Block
+- `Gpt2Embeddings`: token and position embeddings
+- `Gpt2LMHead`: a complete GPT2 model with embedding, transformer, and LM head.
+
+We follow the same breakdown in the implementation of Llama in Levanter.
+
+#### Note on the Implementation Format
+- Each class will have its key layers and components defined as attributes and be initialized with a static method `init()`.
+- Each class will inherit from `StateDictSerializationMixin` from `torch_serialization` and Equinox's `Module` class.
+
+### Mapping weights from/to Hugging Face
+To load weights from Hugging Face, you will need to write a class function `from_hf_state_dict()` in each of your model class. It takes in a Hugging Face state_dict and returns a Levanter state_dict.
+
+For implementation, there are a few helper classes from `torch_serialization` that you can use:
+- To add specific prefix to the keys of Hugging Face state_dict, you can use the helper function `apply_prefix()`. The prefix comes from the name of attributes defined at the beginning of your model class.
+- To unflatten the linear layers of Hugging Face, you can use the helper function `unflatten_linear_params()`.
+- To unstack the transformer blocks of Hugging Face, you can use the helper function `unstack_transformer_blocks()`.
```

Comment:
this isn't really right: `Gpt2MLP` doesn't need custom serialization logic.


```suggestion
### Serialization to/from State Dicts

PyTorch and Hugging Face Transformers use "state dicts" as their preferred serialization format, either as pickles or as the new [safetensors](https://github.com/huggingface/safetensors) format. A state dict is a Python `dict` with string keys and tensor values. The keys of the dict are json-ish "key paths" like `model.blocks.0.mlp.c_proj` and the values are the corresponding parameters for that key path.

You can read more about [PyTorch State Dicts here](https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html).

Levanter has machinery for (de)serializing to and from state dicts. Simple cases are handled automatically, but often times custom logic is needed.

#### Easy Case: Identical Module Structure

If your module has exactly the same fields with the same names and same shapes as the Hugging Face state dict (e.g. Gpt2Mlp), you don't need to do anything.

#### Medium Case: Different Names

If for some reason you want to use different names from the HF implementation (e.g. because the names from HF aren't clear...), you can implement `StateDictSerializationMixin` and use `_state_dict_key_map` to rename keys. For instance, the `Gpt2Transformer` class has this method:

```python
class Gpt2Transformer(StateDictSerializationMixin, eqx.Module):

    ...
    
    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
        return {"blocks": "h"}
```

This says that the field called `blocks` in this class should be (de)serialized as `h`, because the HF GPT-2 implementation uses `h`, which in our opinion isn't very clear. You can also "flatten" the submodules of a field by using `None` or even include `.`s in the name if needed.

#### Hard Case: Custom Serialization

If your modules need special logic, you'll need to implement class `StateDictSerializationMixin` and override `to_state_dict` and `from_state_dict`. It takes in a Hugging Face state dict and returns a Levanter state_dict.

For implementation, there are a few helper classes from `torch_serialization` that you can use:
- To add specific prefix to the keys of a state dict, you can use the helper function `apply_prefix()`. The prefix comes from the name of attributes defined at the beginning of your model class.
- To unflatten the linear layers of a state dict, you can use the helper function `unflatten_linear_params()`.
- To unstack the transformer blocks of a state dict, you can use the helper function `unstack_transformer_blocks()`.
```